### PR TITLE
test: initial support for property-based testing using hypothesis

### PR DESCRIPTION
**Pull request by kszucs**
Still need to configure a CI build for it.

### CODE REVIEW

File: ibis/tests/expr/test_schema.py [hunk 1]
```diff
@@ -134,6 +136,14 @@ def test_whole_schema():
     assert customers.schema() == expected
 
 
+def test_schema_names_and_types_legth_must_match():
```
Comment (cpcloud) [inline ibis/tests/expr/test_schema.py:139]:
```suggestion
def test_schema_names_and_types_length_must_match():
```

File: ibis/tests/strategies.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    )
+)
+
+
+# Strategies for generating schema
+
+
+@st.composite
+def schema(draw, item_strategy=primitive_dtypes, max_size=20):
+    num_fields = draw(st.integers(min_value=0, max_value=max_size))
```
Comment (cpcloud) [inline ibis/tests/strategies.py:165]: Should we still keep an empty schema as a valid construction?

Comment (kszucs) [reply ibis/tests/strategies.py:165]: If an empty table is valid then an empty schema should be valid too.

Comment (cpcloud) [reply ibis/tests/strategies.py:165]: How about raising an exception on empty schema construction and disallowing them altogether?

Comment (kszucs) [reply ibis/tests/strategies.py:165]: I wouldn't disallow empty schema construction, it is also allowed by pyarrow.

**Comment by kszucs** [comment]
Additional work to do, preferably in follow-up PRs:
 - configure a CI job running hypothesis builds in a reproducible fashion
 - utilize the hypothesis strategies for more test cases

@cpcloud can we merge this and handle the rest in follow-ups?

**Comment by cpcloud** [comment]
@kszucs Yeah, that's fine. Can you put an upper bound on the data generation for the test that's failing? The others that passed are likely just barely doing so.

**Comment by kszucs** [comment]
Limited to generate memtables with 5 columns at most.

**Review by cpcloud — APPROVED** [review]

**Pull request merged**